### ISSUE TITLE
[UWP] LoadJournal FileLoadException

### ISSUE THREAD

**Issue by SeriousGeorge** [markmeeus/MarcelloDB#78]
**Plattform: UWP
MarcelloDB Version: 1.08**

[EDIT] Oh.. I just saw, there's 1.09 now which fixes a race-condition in the collections. Maybe my problem has something to do with it?

I'm using MarcelloDB for a while now. But in one project I'm getting a lot of error/crash reports in MS Appcenter which I can't reproduce. The error indicates the method "Journal.LoadJournal()"

It seems there's a race- or fileaccess-condition.
I'm initializing Marcello right at app start in the app.cs.
Sometimes the error happens right at start and sometimes only after some time at user triggered events where db access is executed.

The latter I got fixed I think, because I guess I had two sessions at the same time.. I moved all the logic into the same class. 
But for the crash right at the beginning... I don't know what the problem is.

**Here's a bit of info from the crashlog**
`System.Runtime.ExceptionServices
ExceptionDispatchInfo.Throw () f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\System\Runtime\ExceptionServices\ExceptionDispatchInfo.cs at 63:13
System.Runtime.CompilerServices
TaskAwaiter.ThrowForNonSuccess (Task task) f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\shared\System\Runtime\CompilerServices\TaskAwaiter.cs at 186:25
System.Runtime.CompilerServices
TaskAwaiter.HandleNonSuccessAndDebuggerNotification (Task task) f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\shared\System\Runtime\CompilerServices\TaskAwaiter.cs at 154:48
System.Runtime.CompilerServices
TaskAwaiter.ValidateEnd (Task task) f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\shared\System\Runtime\CompilerServices\TaskAwaiter.cs at 125:17
System.IO.WindowsRuntimeStorageExtensions
<OpenStreamForWriteAsyncCore>d__3.MoveNext ()
System.Runtime.ExceptionServices
ExceptionDispatchInfo.Throw () f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\System\Runtime\ExceptionServices\ExceptionDispatchInfo.cs at 63:13
System.Threading.Tasks
Task.ThrowIfExceptional (Boolean includeTaskCanceledExceptions) f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\System\Threading\Tasks\Task.cs at 2027:17
System.Threading.Tasks
Task.Wait (Int32 millisecondsTimeout, CancellationToken cancellationToken) f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\System\Threading\Tasks\Task.cs at 2757:17
System.Threading.Tasks
Task.Wait () f:\dd\ndp\fxcore\CoreRT\src\System.Private.CoreLib\src\System\Threading\Tasks\Task.cs at 2623:13
MarcelloDB.uwp.FileStorageStream
MarcelloDB.uwp.FileStorageStream..ctor(StorageFolder, String) + 0x38
MarcelloDB.uwp
FileStorageStreamProvider.GetStream (String)
MarcelloDB.Storage.StreamActors
Reader.Read (Int64, Int32)
MarcelloDB.Transactions
**Journal.LoadJournal ()**
MarcelloDB.Transactions
Journal.Apply ()
MarcelloDB
Session.EnsureTransaction ()
MarcelloDB
Session.Transaction (Action)
MarcelloDB.Collections.CollectionFile.Collection[T, TID](String, Func2) + 0x2ee`

**And that's the way I init Marcello at onLaunched in app.xaml.cs through a helper class:**
`  
 StorageFolder DbFolder = await LocalFolder.CreateFolderAsync("db", CreationCollisionOption.OpenIfExists); 
MarcelloDB.uwp.Platform platform = new MarcelloDB.uwp.Platform();
Session = new Session(platform, DbFolder.Path);
CourseFile = Session["courses.dat"];
WeekFile = Session["weeks.dat"];
CourseCollection = CourseFile.Collection<CourseItem, Guid>("course", g => g.guid);
WeekCollection = WeekFile.Collection<WeekItem, Guid>("week", g => g.guid);`

I provoced the error manually with blocking the journal file - and the error surfaced at the first collection access. So my guess is thats also the problem under "real life" conditions. Tough I can't reproduce the error without manipulating/forcing the journal file to be locked manually for testing.

**Comment by SeriousGeorge** [comment]
Just a short followup... even with Update to 1.09 and optimizing the session Usage combined with ErrorHandling we still receive many FileLoadException in Journal.LoadJournal(). Right at the app start.

Any news or advices on this?

**Comment by markmeeus** [comment]
Hi George,

Sorry for the late reply, I’m on holiday atm.

It definitely looks like the journal file got locked somehow. This will for sure happen when you oprn multiple sessions in the same folder.

Is it possible that your app runs in multiple instances? Or that your session creating code is invoked on multiple threads in parallel?
The latter you could solve by placing a lock around the code, the former would be more difficult because MarcelloDb does not support multiple process acces in the same db.

**Comment by SeriousGeorge** [comment]
Hi Mark,

don't meant to disturb your holidays ;)

**Just as FYI at the moment:**
No, the app is a UWP app and has no multi-instance capabilities. I can't even reproduce. But some users are somehow able to trigger that exception.

The Session Creation code is also on a single point on app.xaml.cs - which is exactly triggered once.

**Lock / Semaphore**
But I took your advice nevertheless and wrapped the creation code in a lock - plus, putting the execution of the creation class - since it is async - in the app.xaml.cs in a semaphore (https://blog.cdemi.io/async-waiting-inside-c-sharp-locks/) 

So the execution part in app.xaml.cs looks now like this:

```
 await semaphoreSlim.WaitAsync();

            try
            {
                await Database.InitDb();
            }
            finally
            {
                semaphoreSlim.Release();
            }
```

and the initDb() Class now looks like this:

```
 public static async Task InitDb()
        {
            StorageFolder DbFolder = await LocalFolder.CreateFolderAsync("db", CreationCollisionOption.OpenIfExists);
        
            MarcelloDB.uwp.Platform platform = new MarcelloDB.uwp.Platform();
            Session = new Session(platform, DbFolder.Path);
            CourseFile = Session["courses.dat"];
            WeekFile = Session["weeks.dat"];

          lock (lockObject)
            {
                try
                {
                    CourseCollection = CourseFile.Collection<CourseItem, Guid>("course", g => g.guid);
                    WeekCollection = WeekFile.Collection<WeekItem, Guid>("week", g => g.guid);
                }
...
            }
```

Have to wait until the new version spreads an enough analytics data is available.

**Comment by markmeeus** [comment]
Hi George,

Your code will now make sure that creating the 2 collections will be serialized. This will probably not fix your problem. 
MarcelloDb keeps a journal file to write it’s transactions for the entire session, and the error is thrown because this file is locked.


You may want to protect the InitDb method from being executed twice, in a thread safe way.

```
lock(lockObject) {
  if(Session==null){
     ... create session and collections
  }
}
```

This being said, I found this ticket:
https://github.com/felixrieseberg/electron-windows-store/issues/30

There once was a windows version which kept files locked after the app was closed. So it might be users with an older Windows version?

A healthy approach would be to call Dispose on the Session when the app closes. This will close all underlying filestreams explicitly.

**Comment by SeriousGeorge** [comment]
Hi Mark,

thank you for your suggestions. 
I implemented it the way you described and so far so good. It seems there are no longer FileLoadExceptions. 

On a sidenote tough... I skipped/removed the session dispose in app closing since I got several "ObjectDisposed" Exceptions from several users within the same app lifecycle (dont know whats going on there). Since I removed the session dispose the logs kept quiet.

Thanks for your help
